File: core/src/com/colorflow/entity/dot/DotPool.java
package com.colorflow.entity.dot;

import com.badlogic.gdx.utils.Pool;
import com.colorflow.entity.Path;
import com.colorflow.utility.Position;

/**
 * Created by daniele on 03/05/17.
 */

public class DotPool extends Pool<Dot> {

    private static DotPool instance = new DotPool();

    private DotPool() {
    }

    public static DotPool getInstance() {
        return instance;
    }

    @Override
    protected Dot newObject() {
        return new Dot();
    }

    public Dot get(Dot.Type type, Color color, Path.Type pathType, Position.Radial start, float speed) {
        Dot d = obtain();
        d.set(type, color, pathType, start, speed);
        return d;
    }
}

File: core/src/com/colorflow/play/entity/dot/DotPool.kt
package com.colorflow.play.entity.dot

import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.utils.Pool
import com.colorflow.play.entity.Path
import com.colorflow.AssetProvider
import com.colorflow.graphic.Color
import com.colorflow.graphic.Position

class DotPool(protected val assets: AssetProvider) : Pool<Dot>() {
    protected val used_dots: ArrayList<Dot> = ArrayList()

    override fun newObject(): Dot {
        return Dot(assets, this)
    }

    fun get(type: Dot.Type, color: Color, pathType: Path.Type, start: Position.Radial, speed: Float): Dot {
        val d = obtain()
        d.set(type, color, pathType, start, speed)
        used_dots.add(d)
        return d
    }

    override fun free(`object`: Dot?) {
        used_dots.remove(`object`)
        super.free(`object`)
    }

    fun destroy_all(cb: (Dot)->Unit = {}) {
        used_dots.forEach { dot ->
            cb(dot)
            dot.addAction(Actions.run { dot.destroy {} })
        }
        used_dots.clear()
    }
}
